### ISSUE TITLE
[DeepSeek V4 Pro] Phase 1 correctness divergence: 1.07 max_abs_diff (compiled vs eager) at gb=0

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#108]
## Summary

DeepSeek V4 Pro Phase 1 correctness check (#67) found a **1.07 max_abs_diff** between compiled and eager output — 4 orders of magnitude above the 1e-4 tolerance from `pytorch/benchmarks/dynamo/common.py:1069`. **Eager is verified byte-deterministic** (`eager_self_diff = 0.0`, `eager_deterministic = true`), so this is NOT RNG noise — torch.compile is producing a meaningfully different output from eager on this model.

## Reproducer

```bash
# 1. Ensure transformers V4 PR branch is at /tmp/transformers-v4 (clone PR #45643 sha a0a8482927a1ef14b1e62cd4d932e9bc29a2cd77)
# 2. Run via the generic eval runner
PYTHONPATH=/tmp/transformers-v4/src ~/envs/torch211/bin/python \
    experiments/scripts/run_eval.py \
    --config experiments/configs/deepseek-v4-pro-phase1.json
```

Phase 1 config: tiny (vocab=4096, hidden=7168, layers=4, experts=16, moe_intermediate=3072, batch=1, seq=16). All V4 architectural features active: MLA, hash_layers=3, MCH (Manifold-Constrained Hyper-Connections), MoE routing, hyper-connections.

## What we know

- **gb=0**: model fully captures into a single graph (1712 ops). The divergence is NOT at a graph-break boundary — it's inside Inductor / AOTAutograd's transformations.
- **Eager is deterministic**: 0.0 self-diff across multiple eager runs with corpus's RNG patching applied. Rules out dropout/sampling noise.
- **bf16 + TF32 high precision** — both eager and compiled use the same dtype configuration, so this isn't a precision-mismatch artifact.
- **3.58× speedup** with the same compiled output — perf is fine, the COMPUTATION is wrong.

## Triage paths to try

1. **Backend bisect**: re-run with `aot_eager` (Dynamo + AOTAutograd, no Inductor) instead of `inductor`. If `aot_eager` matches eager bitwise, the divergence is in Inductor. If it diverges too, it's in AOTAutograd or a Dynamo decomposition.
2. **Layer bisect**: hook into model's intermediate outputs (per-block hidden_states), find the first layer where compiled vs eager diverges. Narrows the responsible op.
3. **MoE-specific**: DeepSeek's MoE dispatch is unusual (hash routing + sinkhorn iterations + custom scoring `sqrtsoftplus`). Likely candidate for the divergence — try with `n_routed_experts=1` (no routing) to isolate.
4. **Hyper-connections specific**: MCH iterates `hc_sinkhorn_iters=20` and uses `hc_eps=1e-06`. If Inductor reorders or fuses the sinkhorn loop, accumulated rounding could explain 1.07 max_abs_diff.

## Tested On

- PyTorch `2.11.0+cu128` (run date 2026-04-25)
- transformers `5.7.0.dev0` (PR #45643 branch)
- 1× H100 80GB
- bf16, TF32 high precision
- Phase 1 config (see plan)

## Why this matters

DeepSeek V4 Pro is a high-profile recent release. Community/users will compile-trace this model and may not check correctness as carefully as Phase 1 did. A 1.07 absolute divergence on logits could affect downstream sampling significantly. Worth understanding the root cause before users hit it in production.

## Phase 2 dependency

The original plan gates Phase 2 (full perf at scale) on Phase 1 correctness sign-off. Phase 2 is BLOCKED until this divergence is investigated and either:
- Resolved (Inductor/Dynamo fix lands)
- Acknowledged as expected (e.g., MoE routing inherently non-bitwise; Phase 2 proceeds with documented caveat)

## References

- Plan: [`experiments/deepseek_v4_pro_eval_plan.md`](https://github.com/penguinwu/oss-model-graph-break-corpus/blob/main/experiments/deepseek_v4_pro_eval_plan.md)
- Phase 1 README: `experiments/results/deepseek_v4_pro/phase1-tiny-20260425-183222/README.md`
- Phase 1 results.json: same dir
- Umbrella: #66

**Comment by penguinwu** [comment]
**Update — bf16-specific divergence; fp32 is bitwise-equal (2026-05-02)**

Re-ran DSV4 correctness via the corpus's standard path tonight (post-merge transformers, post-cleanup of one-off `run_eval.py`). With the standard path's default fp32 dtype:

- `max_diff = 0.0`
- `bitwise_equal = true`
- `status = match`

The Apr 25 `max_diff=1.07` finding loaded the model with `dtype=bfloat16` explicitly. So the divergence is **bf16-specific** — most likely bf16 accumulation noise rather than a torch.compile correctness bug.

**Why this distinction matters:** the corpus's standard tolerance check (`tools/file_issues.py` correctness pipeline) uses fp32 by default per upstream `pytorch/benchmarks/dynamo/common.py:1069`. The bf16 case is real-world relevant (DSV4's published Pro/Flash variants ship in FP4+FP8 mixed precision, much lower than bf16) but it's a separate question — "is bf16 numerics stable under torch.compile" applies to many models, not just DSV4.

**Closing this issue as not-a-DSV4-specific-bug.** A separate, broader investigation into bf16 numerical stability under torch.compile may be warranted but isn't DSV4-scoped. If you want a bf16 sweep, file as a new corpus tooling/methodology issue (e.g., `[corpus] add bf16 dtype mode to correctness sweep`).
